### PR TITLE
Migrate to docusaurus

### PR DESCRIPTION
**Pull request by ibraheemdev**
image
image

- [x] re-add some files (license, cname)
- [x] replace shortcodes
- [ ] update home page

### CODE REVIEW

**Comment by nikstur** [comment]
Is there (still) interest by the maintainers in migrating to Docusaurus? I don't know in what detail this has been discussed before, but here is my opinion:

Im generally in favour of Docusaurus:

Pros:
- Designed solely for documentation
- Easier to support multiple versions of the website for different releases of actix-web (there is no straightforward way in Hugo)
- Easier to support i18n (less manual work needed than in Hugo)
- Looks prettier (including night mode --- very subjective of course) 
- Benefit from updates through a simple version bump instead of building it by ourselves
- Solves license issue mentioned in the readme of this repo.

Cons:
- Variable substitution in markdown is a bit ugly. The entire line has to be wrapped in an JSX element.  
  
  Example:
  ```jsx
  <>Actix Web currently has a minimum supported Rust version (MSRV) of { rustVersion }</>
  ```
  
  It's even worse in code blocks: The entire block has to be written in JSX and the content has to be escaped to retain line breaks.
  ```jsx
  <RenderCodeBlock className="language-toml">
  {`[dependencies]
  actix-web = "${actixWebVersion}"`}
  </RenderCodeBlock>
  ```
  However right now variables are not used very often: the rust and actix-web versions are mentioned in just two files each. It might be feasible to either hardcode the variables or use the "ugly" JSX syntax 
- Javascript tooling (npm or yarn)
- Yet another tool to learn and support (however its still mostly markdown of course)

If there is interest, I'd volunteer for reviewing the PR in depth (I've looked over it but haven't commented on the issues I found) and helping out with the finishing touches if @ibraheemdev is not available anymore.

**Comment by robjtede** [comment]
From the maintainers, there is still interest in migrating. The only thing holding this PR back is a homepage refresh.

**Comment by ibraheemdev** [comment]
> I'd volunteer for reviewing the PR in depth (I've looked over it but haven't commented on the issues I found) and helping out with the finishing touches

That would be great :)

> The only thing holding this PR back is a homepage refresh.

I think to avoid the list of conflicts growing even more, we can use the current home page and refresh it later.

File: examples/flexible-responders/src/main.rs [hunk 1]
```diff
@@ -14,6 +14,7 @@ async fn hello_world() -> impl Responder {
 async fn current_temperature() -> impl Responder {
     web::Json(Measurement { temperature: 42.3 })
 }
+// </flexible-responders>
```
Comment (nikstur) [inline examples/flexible-responders/src/main.rs:17]: Same here as in examples/easy-form-handling/src/main.rs

File: examples/powerful-extractors/src/main.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 use actix_web::{web, App, HttpResponse, HttpServer, Responder};
 use serde::{Deserialize, Serialize};
 
+// <powerful-extractors>
```
Comment (nikstur) [inline examples/powerful-extractors/src/main.rs:4]: Same as in examples/easy-form-handling/src/main.rs

File: examples/request-routing/src/main.rs [hunk 3]
```diff
@@ -1,6 +1,7 @@
-// <request-routing>
```
Comment (nikstur) [inline examples/request-routing/src/main.rs:1]: Same as in examples/easy-form-handling/src/main.rs

File: package.json [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  },
+  "dependencies": {
+    "@docusaurus/core": "2.0.0-alpha.72",
+    "@docusaurus/preset-classic": "2.0.0-alpha.72",
+    "@fortawesome/fontawesome-svg-core": "^1.2.35",
+    "@fortawesome/free-solid-svg-icons": "^5.15.3",
+    "@fortawesome/react-fontawesome": "^0.1.14",
+    "@mdx-js/react": "^1.6.21",
+    "clsx": "^1.1.1",
+    "docusaurus-plugin-sass": "^0.1.12",
```
Comment (nikstur) [inline package.json:24]: We should probably update all dependencies, but we definiteley need to update `docusaurus-plugin-sass` and add `sass` as a dev dependency.
The current version pulls in a node-sass version that requires Python2 (and thus will fail on many modern systems).

File: sidebars.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        'Introduction': [
+            'welcome',
+            'whatis'
+        ],
+        'Basics': [
+            'getting-started',
+            'application',
+            'server',
+            'extractors',
+            'handlers',
```
Comment (nikstur) [inline sidebars.js:12]: These are the exchanged in the current docs. The handlers page comes first

File: src/pages/index.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
+import clsx from "clsx";
+import React from "react";
+import Link from "@docusaurus/Link";
+import Layout from "@theme/Layout";
+import Tabs from "@theme/Tabs";
+import TabItem from "@theme/TabItem";
```
Comment (nikstur) [inline src/pages/index.js:6]: These are unused and can be removed.

File: .gitignore [hunk 7]
```diff
@@ -1,7 +1,4 @@
-public
-tmp
-_site
-Cargo.lock
```
Comment (nikstur) [inline .gitignore:4]: Is this what we want or is it accidental? If it was accidental, the committed Cargo.lock file should also be removed.

**Review by nikstur — CHANGES_REQUESTED** [review]
LGTM, however there are a few (larger) issues we should fix before merging:

- [ ] Fix conflicts 
- [ ] Update `docusaurus-plugin-sass` and add `sass` dev dependency
<s>- [ ] Factor out changes to examples (actual content) into a separate PR</s>

A few things we can and should do after merging in separate PRs:

- Remove superfluous information from yaml metadata header in all markdown files
- Fix headings: Hugo renders them correctly (h1 turns into h2), however, docusaurus doesn't, so they all need to be h2 headings

**Comment by Woile** [comment]
@ibraheemdev I think I can finish this PR, do you mind if I continue it?

**Comment by ibraheemdev** [comment]
Sure! I don't think I have the energy to fix all the merge conflicts 😅

**Comment by robjtede** [comment]
Surely easier to re-apply the changes from this diff manually in its own fork. There's been a lot of changes and I wouldn't trust a merge conflict solve at this point.

File: docs/getting-started.md [hunk 8]
```diff
@@ -4,14 +4,15 @@ menu: docs_basics
...
+
 ## Installing Rust
 
 If you don't have Rust yet, we recommend you use `rustup` to manage your Rust installation. The
 [official rust guide][rustguide] has a wonderful section on getting started.
 
-Actix Web currently has a minimum supported Rust version (MSRV) of {{< rust-version "actix-web" >}}.
-Running `rustup update` will ensure you have the latest and greatest Rust version available. As
-such, this guide assumes you are running Rust {{< rust-version "actix-web" >}} or later.
+<>Actix Web currently has a minimum supported Rust version (MSRV) of { rustVersion }. Running <code>rustup update</code> will ensure you have the latest and greatest Rust version available. As such, this guide assumes you are running Rust { rustVersion } or later.</>
```
[the remark this answers is not in the document]
Comment (Woile) [reply docs/getting-started.md:15]: For #266 using `p` which makes it same as a markdown text.

File: docs/getting-started.md [hunk 9]
```diff
@@ -24,16 +25,19 @@ cd hello-world
...
 Add `actix-web` as a dependency of your project by adding the following to your `Cargo.toml` file.
 
+
+<!-- DEPENDENCY -->
+
 ```toml
 [dependencies]
-actix-web = "{{< actix-version "actix-web" >}}"
+actix-web = "3"
 ```
```
[the remark this answers is not in the document]
Comment (Woile) [reply docs/getting-started.md:34]: done 4 #266

File: docs/server.md [hunk 10]
```diff
@@ -84,13 +86,15 @@ for this very reason, and they should also be implemented asynchronously where n
...
 `rustls` integration and `openssl` is for `openssl`.
 
+<!-- DEPENDENCY -->
+
 ```toml
 [dependencies]
-actix-web = { version = "{{< actix-version "actix-web" >}}", features = ["openssl"] }
+actix-web = { version = "3", features = ["openssl"] }
 openssl = { version = "0.10" }
 ```
```
[the remark this answers is not in the document]
Comment (Woile) [reply docs/server.md:95]: Done #266

File: docs/websockets.md [hunk 11]
```diff
@@ -11,7 +11,8 @@ websocket communications with an http actor.
 
 The following is an example of a simple websocket echo server:
 
-{{< include-example example="websockets" file="main.rs" section="websockets" >}}
+import CodeBlock from "../src/components/code_block.js";
```
[the remark this answers is not in the document]
Comment (Woile) [reply docs/websockets.md:14]: Done #266

File: docs/whatis.md [hunk 12]
```diff
@@ -26,8 +28,7 @@ in the complete absence of another HTTP server `actix-web` is powerful enough to
 HTTP/2 support as well as TLS (HTTPS). This makes it useful for building small services ready for
 production.
 
-Most importantly: `actix-web` runs on Rust {{< rust-version "actix-web" >}} or later and it works
-with stable releases.
+<>Most importantly: <code>actix-web</code> runs on Rust { rustVersion } or later and it works with stable releases.</>
```
[the remark this answers is not in the document]
Comment (Woile) [reply docs/whatis.md:31]: Done https://github.com/actix/actix-website/pull/266

**Comment by Woile** [comment]
I think I've covered most of the comments in the PR #266 can u take a look @nikstur ? Thanks

**Comment by nikstur** [comment]
Great, thank you for the work! I'll take a look.

**Comment by robjtede** [comment]
superseded by #266

**Pull request closed without merging**